core/src/main/scala/diffson/jsonpatch/JsonDiff.scala

```diff
@@ -20,73 +20,45 @@ package jsonpatch
-    if (json1 === json2)
-      // if they are equal, this one is easy...
-      Nil
-    else
-      (json1, json2) match {
-        case (JsObject(fields1), JsObject(fields2))      => fieldsDiff(fields1.toList, fields2.toList, pointer)
-        case (JsArray(arr1), JsArray(arr2)) if diffArray => arraysDiff(arr1.toList, arr2.toList, pointer)
-        case (_, _) => List(Replace(pointer, json2, if (rememberOld) Some(json1) else None))
-      }
+  private def diff(json1: Json, json2: Json, pointer: Pointer): Eval[Chain[Operation[Json]]] =
+    (json1, json2) match {
+      case (JsObject(fields1), JsObject(fields2))      => fieldsDiff(fields1.toList, fields2, pointer)
+      case (JsArray(arr1), JsArray(arr2)) if diffArray => arraysDiff(arr1.toList, arr2.toList, pointer)
+      case _ if json1 === json2                        =>
+        // if they are equal, this one is easy...
+        Eval.now(Chain.empty)
+      case _ => Eval.now(Chain.one(Replace(pointer, json2, if (rememberOld) Some(json1) else None)))
+    }
 
   private def fieldsDiff(fields1: List[(String, Json)],
-                         fields2: List[(String, Json)],
-                         path: Pointer): List[Operation[Json]] = {
-    // sort fields by name in both objects
-    val sorted1 = fields1.sortBy(_._1)
-    val sorted2 = fields2.sortBy(_._1)
-    @tailrec
-    def associate(fields1: List[(String, Json)],
-                  fields2: List[(String, Json)],
-                  acc: List[(Option[(String, Json)], Option[(String, Json)])])
-        : List[(Option[(String, Json)], Option[(String, Json)])] = (fields1, fields2) match {
-      case (f1 :: t1, f2 :: t2) if f1._1 == f2._1 =>
-        // same name, associate both
-        associate(t1, t2, (Some(f1), Some(f2)) :: acc)
-      case (f1 :: t1, f2 :: _) if f1._1 < f2._1 =>
-        // the first field is not present in the second object
-        associate(t1, fields2, (Some(f1), None) :: acc)
-      case (_ :: _, f2 :: t2) =>
-        // the second field is not present in the first object
-        associate(fields1, t2, (None, Some(f2)) :: acc)
-      case (_, Nil) =>
-        fields1.map(Some(_) -> None) ::: acc
-      case (Nil, _) =>
-        fields2.map(None -> Some(_)) ::: acc
+                         fields2: Map[String, Json],
+                         path: Pointer): Eval[Chain[Operation[Json]]] =
+    fields1 match {
+      case (fld, value1) :: fields1 =>
+        fields2.get(fld) match {
+          case Some(value2) =>
+            fieldsDiff(fields1, fields2 - fld, path).flatMap(d => diff(value1, value2, path / fld).map(_ ++ d))
+          case None =>
+            // field is not in the second object, delete it
+            fieldsDiff(fields1, fields2, path).map(
+              _.prepend(Remove(path / fld, if (rememberOld) Some(value1) else None)))
+        }
+      case Nil =>
+        Eval.now(Chain.fromSeq(fields2.toList).map { case (fld, value) => Add(path / fld, value) })
```

Comment:
```suggestion
        Eval.now(Chain.fromSeq(fields2.toList.map { case (fld, value) => Add(path / fld, value) }))
```
I'm pretty sure `map` is slightly faster on `List`. I also wonder whether one can go more directly from `Map` to `Chain`, but you probably tried already.